### PR TITLE
`.write_data_to_vtk()` method in meshes

### PR DESCRIPTION
**Pull request by RemDelaporteMathurin**
Hi all,

This is my first _real_ contribution to OpenMC 🎉 

**Current**
I have been using `CylindricalMesh` filters for a while now (with @shimwell on fusion applications), and thought tallies using it could use a "to vtk" method to avoid external repetitive scripting.
Moreover, the `openmc-voxel-to-vtk` script seems to only handle `RegularMesh`.

**New behaviour**
This PR adds a `write_data_to_vtk` method to the classes `RegularMesh,` `CylindricalMesh` and `RectilinearMesh` and `SphericalMesh` on the similar structure to `UnstructuredMesh.write_data_to_vtk` (thanks @pshriwise for the tip).

**Usage**

```python
import openmc
import numpy as np

cylinder_mesh = openmc.CylindricalMesh()
cylinder_mesh.r_grid = np.linspace(1, 2, num=30)
cylinder_mesh.phi_grid = np.linspace(0, np.pi, num=50)
cylinder_mesh.z_grid = np.linspace(0, 1, num=30)

data = np.random.random(
    (cylinder_mesh.dimension[0]*cylinder_mesh.dimension[1]*cylinder_mesh.dimension[2])
)
cylinder_mesh.write_data_to_vtk(
    filename="cylindrical.vtk",
    datasets={"mean": data, "std_dev": data},
    volume_normalization=False,
)
```

cylindrical.vtk
image

Equivalent for `SphericalMesh`

image

**Limitations**
- For tallies with `CylindricalMesh` or `SphericalMesh` filters, the angular discretisation doesn't look great with too few elements. @pshriwise seems to have a solution for that 💪 

image

### CODE REVIEW

**Comment by shimwell** [comment]
Naturally I'm convinced by this addition but just wanted to mention ```vtk``` is an optional dependency at the moment https://github.com/openmc-dev/openmc/blob/d7a456155f1bc4bc38c0130103a406405e9d6840/setup.py#L77

**Comment by RemDelaporteMathurin** [comment]
> Naturally I'm convinced by this addition but just wanted to mention ```vtk``` is an optional dependency at the moment, so I guess some of the CI will fail due to the ```import vtk``` https://github.com/openmc-dev/openmc/blob/d7a456155f1bc4bc38c0130103a406405e9d6840/setup.py#L77

Ah good point! Surely I can add the import statement in the method itself!

File: openmc/tallies.py
Comment (shimwell) [inline openmc/tallies.py:12]: might break the CI due to vtk being an optional dependency
```suggestion
```

File: openmc/tallies.py
Comment (shimwell) [inline openmc/tallies.py:473]: Perhaps something like this

```suggestion
        try:
                import vtk
                import vtk.util.numpy_support as nps
        except ImportError:
                msg = 'Python package vtk was not found, please install vtk to use Tally.write_to_vtk.'
                raise ImportError(msg)
        mesh = None
```

File: tests/unit_tests/test_mesh_to_vtk.py
Comment (paulromano) [inline tests/unit_tests/test_mesh_to_vtk.py:41]: If you use `pathlib.Path`, this can become
```suggestion
    assert filename.is_file()
```

File: tests/unit_tests/test_mesh_to_vtk.py
Comment (paulromano) [inline tests/unit_tests/test_mesh_to_vtk.py:5]: You can use [`pytest.importorskip`](https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest-importorskip) so that  these tests will get skipped if vtk is not available

**Review by paulromano — CHANGES_REQUESTED** [review]
Thanks for the contribution and congratulations on your first PR! :tada: As we discussed offline, a little refactoring could help cut down on the code duplication. Other than that, I have a few minor comments that should be addressed but I don't see any major issue that will stop this from getting merged. Nice work including documentation and tests!

File: tests/unit_tests/test_mesh_to_vtk.py
Comment (paulromano) [reply tests/unit_tests/test_mesh_to_vtk.py]: Ah, you're right! I thought it was on `StructuredMesh` but I stand corrected. Yes, it would be good to have a separate PR to make it consistent across the mesh classes :+1:

File: openmc/mesh.py
Comment (RemDelaporteMathurin) [reply openmc/mesh.py:706]: I don't think we can do x_vals, y_vals and z_vals cause some of these get converted. But we can pass the points all at once. On it

**Comment by RemDelaporteMathurin** [comment]
@paulromano thanks for the review! I made the requested changes, the tests pass locally.

File: openmc/mesh.py
Comment (paulromano) [inline openmc/mesh.py:211]: The default shows up in the argument list so probably not necessary to list it explicitly here
```suggestion
```

File: openmc/mesh.py
Comment (paulromano) [inline openmc/mesh.py:732]:
```suggestion
        ll, ur = self.lower_left, self.upper_right
        x_vals = np.linspace(ll[0], ur[0], num=self.dimension[0] + 1)
        y_vals = np.linspace(ll[1], ur[1], num=self.dimension[1] + 1)
        z_vals = np.linspace(ll[2], ur[2], num=self.dimension[2] + 1)
```

File: openmc/mesh.py
Comment (paulromano) [inline openmc/mesh.py:1410]: This line is getting a bit long and could use some wrapping over multiple lines

**Review by paulromano — CHANGES_REQUESTED** [review]
A few more small fixes/suggestions

**Review by paulromano — APPROVED** [review]
All looks good now! Thanks for going through the requested changes. In case you're not aware, we have a minimum [36 hour review period](https://docs.openmc.org/en/latest/devguide/contributing.html#contribution-process) to give all community members a chance to chime in before merging, so I'll hold off until we reach that point in case others have requests.

**Comment by pshriwise** [comment]
> Does anyone else think it would be worth adding an argument to the ``write_to_vtk``` method that allows the data to be scaled.
> 
> Often I'm writing out raw tally data (score per source particle) to a vtk file when I actually want data with some units (e.g. Joules for heating). I have the scaling numbers elsewhere in my python script and would like to make use of it when exporting the VTK files.
> 
> We have `volume_normalization` which is nice but could we also have `scaling_factor`?

I agree that one will often want to apply some kind of scaling/normalization factor to tally data. I voted to make these methods of the mesh class rather than the tally class for this reason. I think that these factors can simply be applied to the datasets being written beforehand rather than passing the scaling factor into the function itself, however. I’m thinking that will more flexible than passing a single scaling factor into the function.

**Comment by RemDelaporteMathurin** [comment]
I guess wheter or not to include this comes down to how many lines it saves?

The current use case is something like:

```python
scaling_factor = 10

mesh1.write_data_to_vtk(
    filename="out1.vtk",
    datasets={"mean": data1 * scaling_factor}
)

mesh2.write_data_to_vtk(
    filename="out2.vtk",
    datasets={"mean": data2 * scaling_factor}
)

```

What Jon suggests would be:

```python
scaling_factor = 10

mesh1.write_data_to_vtk(
    filename="out1.vtk",
    datasets={"mean": data1},
    scaling_factor=scaling_factor
)

mesh2.write_data_to_vtk(
    filename="out2.vtk",
    datasets={"mean": data2},
    scaling_factor=scaling_factor
)

```

I'm happy with both as it's just a simple multiplication (in contrast with volume normalisation which requires to get the voxels volumes). I also agree with Patrick on the flexibility aspect of option 1.

**Comment by shimwell** [comment]
I think having data scaling and volume scaling separately could get a bit confusing for the user in some cases.

For example if a user wants the vtk to be in watts per m3

```
mesh.write_data_to_vtk(
    filename="out.vtk",
    datasets={"mean": data * source.strength*eV_to_joules_factor*cm_to_m_factor} # converts from per source particle to watts and from cm3 to meters3
    volume_normalization=True # divides each scaled dataset number by the tet volume
)
```

Perhaps this would be simpler for the user if the ```scaling_factor``` was able to accommodate volume scaling

```
mesh.write_data_to_vtk(
    filename="out.vtk",
    datasets={"mean": data}
    scaling_factor = source.strength*eV_to_joules_factor*cm_to_m_factor / mesh.volume
)
```

This is arguably more flexible than volume_normalisation=True/False as one could do things like divided by the sqrt(mesh.volume). I can't think of a reason for doing that but currently scaling with volume is divided into to parts and on part can only ever be / by volume. 

Which could eventually be upgraded to supporting some standard units like we do elsewhere in the code (e.g. when setting density in "g/cm3") and provide a really easy user interface.

```
mesh.write_data_to_vtk(
    filename="out.vtk",
    datasets={"mean": data}
    scaling_factor = 'w/m3'
)
```

I don't think we could ever get to this level of user simplification with separate scaling for volume_normalisation and scaling on the data sets. Perhaps I'm just thinking longer term with a vision towards having a really simple interface that supports common units and this is not the time or PR for such ideas.

**Comment by paulromano** [comment]
I tend to agree with @RemDelaporteMathurin's point -- we're not really saving the user much by adding that extra argument and it would be hard to anticipate all possible use cases, so I personally think it's better to leave any sort of scaling (other than volume normalization) to be the responsibility of the user.

**Comment by shimwell** [comment]
I don't think these two scaling (volume and user) can be decoupled in all cases

**Comment by RemDelaporteMathurin** [comment]
> I am not really suggesting adding another argument, rather replacing one as I don't think these two scaling (volume and user) can be decoupled in all cases

But here volume_normalisation is defaults to True, so users don't have to worry about it. Whereas if it's in scaling_factor (defaulting to one I suppose), users need to normalise themselves: that's extra work

**Comment by shimwell** [comment]
> ```scaling_factor``` could default to ```1/mesh.volumes``` which would be more explicit

**Comment by RemDelaporteMathurin** [comment]
> could `scaling_factor` default to `1/mesh.volumes`? Which would be more explicit about the opperation

Maybe it could but in that case, users have to remember to divide by the volumes when passing non-default scaling factor...

**Comment by paulromano** [comment]
Let me make a suggestion here -- this PR simply extends the set of classes which support a `write_data_to_vtk` method and those new methods conform to the existing interface that was already present for `UnstructuredMesh`. If we want to consider changing the argument for scaling/normalization, let's do that as a separate PR / issue so as not to hold up this one (which in my opinion is ready to go). Is that acceptable to you @shimwell?

**Review by shimwell — APPROVED** [review]

**Comment by shimwell** [comment]
I can make a PR / issue for having a single normalisation method instead of have two parts to the normalisation another time.

For now this is an improvement on what we have so I think it is a worthwhile upgrade.

**Comment by shimwell** [comment]
Looks like I've previously tried to get a single scaling factor in to no avail https://github.com/openmc-dev/openmc/pull/1886

**Comment by shimwell** [comment]
36 hours has passed and @paulromano and @pshriwise have approved so LGTM

**Pull request merged**
